src/poetry/repositories/pool.py

```diff
@@ -63,133 +81,61 @@ def add_repository(
-        if repository_name in self._lookup:
-            raise ValueError(f"{repository_name} already added")
-
-        if default:
-            if self.has_default():
-                raise ValueError("Only one repository can be the default")
+        if repository_name in self._repositories:
+            raise ValueError(
+                "A repository with (case-independent) name"
+                f" {repository_name} was already added"
+            )
 
-            self._default = True
-            self._repositories.insert(0, repository)
-            for name in self._lookup:
-                self._lookup[name] += 1
+        if default and self.has_default():
+            raise ValueError("Only one repository can be the default")
 
-            if self._secondary_start_idx is not None:
-                self._secondary_start_idx += 1
-
-            self._lookup[repository_name] = 0
+        if default:
+            repository_priority = RepositoryPriority.DEFAULT
         elif secondary:
-            if self._secondary_start_idx is None:
-                self._secondary_start_idx = len(self._repositories)
-
-            self._repositories.append(repository)
-            self._lookup[repository_name] = len(self._repositories) - 1
+            repository_priority = RepositoryPriority.SECONDARY
         else:
-            self._has_primary_repositories = True
-            if self._secondary_start_idx is None:
-                self._repositories.append(repository)
-                self._lookup[repository_name] = len(self._repositories) - 1
-            else:
-                self._repositories.insert(self._secondary_start_idx, repository)
-
-                for name, idx in self._lookup.items():
-                    if idx < self._secondary_start_idx:
-                        continue
-
-                    self._lookup[name] += 1
-
-                self._lookup[repository_name] = self._secondary_start_idx
-                self._secondary_start_idx += 1
-
+            repository_priority = RepositoryPriority.PRIMARY
+        self._repositories[repository_name] = PrioritizedRepository(
+            repository, repository_priority
+        )
         return self
 
     def remove_repository(self, repository_name: str) -> Pool:
-        if repository_name is not None:
-            repository_name = repository_name.lower()
-
-        idx = self._lookup.get(repository_name)
-        if idx is not None:
-            del self._repositories[idx]
-            del self._lookup[repository_name]
-
-            if idx == 0:
-                self._default = False
-
-            for name in self._lookup:
-                if self._lookup[name] > idx:
-                    self._lookup[name] -= 1
-
-            if (
-                self._secondary_start_idx is not None
-                and self._secondary_start_idx > idx
-            ):
-                self._secondary_start_idx -= 1
-
+        del self._repositories[repository_name.lower()]
```

Comment:
The old function had error checking -- this can now throw an `IndexError` -- we should probably check for existence first since we silently swallowed errors before.

Reply:
(and we should also have a test for deleting a non-extant repository)